### ISSUE TITLE
Janus WebRTC Gateway sometimes reports error resolving mDNS addresses

### ISSUE THREAD

**Issue by leigh-johnson** [bitsy-ai/printnanny-os#137]
## **Describe the bug**

Seeing reports of Janus WebRTC Gateway failing to peer folks broadcasting an mDNS address. I managed to reproduce this on the latest stable build of Chromium for x86_64 Linux `Version 107.0.5304.87 (Official Build) (64-bit)`
```
Nov 07 12:53:52 pn-v0-4-0 janus-gateway[576]: [WARN] [7219519908540579] Error resolving mDNS address (5970090f-2a00-44a9-bc37-7c23c7df0a2f.local): Error resolving <E2><80><9C>5970090f-2a00-44a9-bc37-7c23c7df0a2f.local<E2><80><9D>: Name or service not known
```

## **To Reproduce**

1. After FIRST BOOT, go to http://<hostname>/vision and try to start either the camera stream or one of the demo videos
2. Notice: sdp offer/answer succeeds to the point where `webrtcup` event is emitted
```
set remote sdp OK
janus.ts:278 create answer OK
janus.ts:280 set local sdp OK
janus.ts:213 pc.oniceconnectionstatechange => checking
janus.ts:329 start 1 response sent with status starting
janus.ts:213 pc.oniceconnectionstatechange => connected
janus.ts:153 [7219519908540579] streaming handle event status: {status: 'started', id: 1}
janus.ts:114 webrtcup event
janus.ts:237 track.onmute Event {isTrusted: true, type: 'mute', target: MediaStreamTrack, currentTarget: MediaStreamTrack, eventPhase: 2, …}
janus.ts:237 track.onmute Event {isTrusted: true, type: 'mute', target: MediaStreamTrack, currentTarget: MediaStreamTrack, eventPhase: 2, …}
```
3. HOWEVER, the media stream never gets to the "playing state".  Janus Gateway logs show:
```
Nov 07 12:53:52 pn-v0-4-0 janus-gateway[576]: [WARN] [7219519908540579] Error resolving mDNS address (5970090f-2a00-44a9-bc37-7c23c7df0a2f.local): Error resolving <E2><80><9C>5970090f-2a00-44a9-bc37-7c23c7df0a2f.local<E2><80><9D>: Name or service not known
```

I was wondering where the heck `5970090f-2a00-44a9-bc37-7c23c7df0a2f.local` is coming from, found it in ICE candidate exchange
```
sdpMid: v1, sdpMLineIndex: 0, candidate: candidate:1178812653 1 udp 2113937151 5970090f-2a00-44a9-bc37-7c23c7df0a2f.local 49962 typ host generation 0 ufrag xZ5u network-cost 999
```

Which is then followed by an (unsprirising) STUN host lookup error
```
url: stun:stun.l.google.com:19302
address: [0:0:0:x:x:x:x:x]
port: 60012
host_candidate: [0:0:0:x:x:x:x:x]:60012
error_text: STUN host lookup received error.
error_code: 701
```

## **Expected behavior**
Starting camera/video stream works! 100% of the time. :joy: 

```
-----------------------
Build Info:  |
-----------------------
IMAGE_NAME = printnanny-release-image-raspberrypi4-64-20221107203243
DISTRO_PRETTY = PrintNanny Linux 0.4.3 (Dunite)
VARIANT_NAME = PrintNanny OS Core
VARIANT_ID = core
DISTRO = printnanny
DISTRO_NAME = PrintNanny Linux
DISTRO_VERSION = 0.4.3
DISTRO_CODENAME = Dunite
ID_LIKE = BitsyLinux
HOME_URL = https://printnanny.ai
BUG_REPORT_URL = https://github.com/bitsy-ai/printnanny-os/issues
YOCTO_VERSION = 4.0.2
YOCTO_CODENAME = Langdale
MACHINE = raspberrypi4-64
TUNE_PKGARCH = cortexa72
-----------------------
Layer Revisions:      |
-----------------------
meta              = langdale:6b9db5a99bf0fab47219d4648f4663c4397ac87f
meta-poky         = langdale:6b9db5a99bf0fab47219d4648f4663c4397ac87f
meta-yocto-bsp    = langdale:6b9db5a99bf0fab47219d4648f4663c4397ac87f
meta-raspberrypi  = master:722c51647c75d63e8c6c53ee214c4e3f0a263ed2
meta-oe           = langdale:c5668905a6d8a78fb72c2cbf8b20e91e686ceb86
meta-python       = langdale:c5668905a6d8a78fb72c2cbf8b20e91e686ceb86
meta-multimedia   = langdale:c5668905a6d8a78fb72c2cbf8b20e91e686ceb86
meta-bitsy        = printnanny-dash-0-3-15:875d9d38041f1b1c2132d831af7dfdb4fdade511
meta-neural-network = langdale-compat:c8422a7380ef81f75bb5ca2216210b434c4e7114
meta-printnanny   = printnanny-dash-0-3-15:875d9d38041f1b1c2132d831af7dfdb4fdade511
meta-networking   = langdale:c5668905a6d8a78fb72c2cbf8b20e91e686ceb86
meta-filesystems  = langdale:c5668905a6d8a78fb72c2cbf8b20e91e686ceb86
meta-initramfs    = langdale:c5668905a6d8a78fb72c2cbf8b20e91e686ceb86
meta-webserver    = langdale:c5668905a6d8a78fb72c2cbf8b20e91e686ceb86
meta-swupdate     = master:5e611f44db1b9333342ed61d77a3c3ed7be69c2b
workspace         = langdale:6b9db5a99bf0fab47219d4648f4663c4397ac87f
meta-microcontroller = master:9f396e595b2d639be7f2b494f4a61e48d9495058
```

**Comment by leigh-johnson** [comment]
Some extra :hot_pepper: spice:

1. I can't reproduce this AFTER the first boot. EVERY single subsequence boot works.  :thinking:

For now, I'm telling folks to reboot as a work-around - but I'd like to get this right on first boot! If I can't figure out a root cause fix, let's at least reboot automatically the first time PrintNanny OS comes up.

2. This points to some kind of ordering issue between the point where `cloud-init` changes the system'd hostname from the default (raspberrypi-64) to the user's provided hostname and when `systemd-resolved`, `avahi-daemon`, and `janus-gateway` start.


Things to try:

1) Ensure janus-gateway is always started AFTER avahi-daemon, and restart janus-gateway when avahi is restarted 

2) I *think* the mDNS ICE candidate proposal should **never** be getting to the stage where STUN is used to translate the mDNS address, except in super wonky cases where someone is running a NAT to create multiple virtual LANs. 

Let's see if I can reproduce the same behavior using half-trickle ice, where only the initiator performs candidate generation. We only need full trickle ICE in the cloud webrtc deployment.

**Comment by leigh-johnson** [comment]
Ordering of mDNS broadcast with default hostname -> hostname changes via cloud-init.service -> NTP server connection established -> avahi restarted.

:thinking: Noticed avahi-service doesn't re-sign cookie after NTP changes system clock time.

```
pn-v0-4-0:/home/pi# systemctl status systemd-resolved
* systemd-resolved.service - Network Name Resolution
     Loaded: loaded (/lib/systemd/system/systemd-resolved.service; enabled; vendor preset: enabled)
     Active: active (running) since Tue 2022-05-24 14:32:40 PDT; 5 months 14 days ago
       Docs: man:systemd-resolved.service(8)
             man:org.freedesktop.resolve1(5)
             https://www.freedesktop.org/wiki/Software/systemd/writing-network-configuration-managers
             https://www.freedesktop.org/wiki/Software/systemd/writing-resolver-clients
   Main PID: 316 (systemd-resolve)
     Status: "Processing requests..."
      Tasks: 1 (limit: 9002)
     CGroup: /system.slice/systemd-resolved.service
             `-316 /lib/systemd/systemd-resolved

May 24 14:32:39 raspberrypi4-64 systemd[1]: Starting Network Name Resolution...
May 24 14:32:40 raspberrypi4-64 systemd-resolved[316]: Positive Trust Anchors:
May 24 14:32:40 raspberrypi4-64 systemd-resolved[316]: . IN DS 20326 8 2 e06d44b80b8f1d39a95c0b0d7c65d08458e880409bbc683457104237c7f8ec8d
May 24 14:32:40 raspberrypi4-64 systemd-resolved[316]: Negative trust anchors: home.arpa 10.in-addr.arpa 16.172.in-addr.arpa 17.172.in-addr.arpa 18.172.in-addr.arpa 19.172.in-addr.arpa 20.172.in-addr.arpa 21.172>
May 24 14:32:40 raspberrypi4-64 systemd-resolved[316]: Using system hostname 'raspberrypi4-64'.
May 24 14:32:40 raspberrypi4-64 systemd[1]: Started Network Name Resolution.
May 24 14:33:07 pn-v0-4-0 systemd-resolved[316]: System hostname changed to 'pn-v0-4-0'.
Nov 07 13:48:58 pn-v0-4-0 systemd-resolved[316]: Clock change detected. Flushing caches.
```

```
May 24 14:32:48 raspberrypi4-64 systemd[1]: Starting Avahi mDNS/DNS-SD Stack...
May 24 14:32:48 raspberrypi4-64 avahi-daemon[379]: Found user 'avahi' (UID 994) and group 'avahi' (GID 993).
May 24 14:32:48 raspberrypi4-64 avahi-daemon[379]: Successfully dropped root privileges.
May 24 14:32:48 raspberrypi4-64 avahi-daemon[379]: avahi-daemon 0.8 starting up.
May 24 14:32:49 raspberrypi4-64 avahi-daemon[379]: Successfully called chroot().
May 24 14:32:49 raspberrypi4-64 avahi-daemon[379]: Successfully dropped remaining capabilities.
May 24 14:32:49 raspberrypi4-64 avahi-daemon[379]: No service file found in /etc/avahi/services.
May 24 14:32:49 raspberrypi4-64 avahi-daemon[379]: Joining mDNS multicast group on interface lo.IPv6 with address ::1.
May 24 14:32:49 raspberrypi4-64 avahi-daemon[379]: New relevant interface lo.IPv6 for mDNS.
May 24 14:32:49 raspberrypi4-64 avahi-daemon[379]: Joining mDNS multicast group on interface lo.IPv4 with address 127.0.0.1.
May 24 14:32:49 raspberrypi4-64 avahi-daemon[379]: New relevant interface lo.IPv4 for mDNS.
May 24 14:32:49 raspberrypi4-64 avahi-daemon[379]: Network interface enumeration completed.
May 24 14:32:49 raspberrypi4-64 avahi-daemon[379]: Registering new address record for ::1 on lo.*.
May 24 14:32:49 raspberrypi4-64 avahi-daemon[379]: Registering new address record for 127.0.0.1 on lo.IPv4.
May 24 14:32:49 raspberrypi4-64 systemd[1]: Started Avahi mDNS/DNS-SD Stack.
May 24 14:32:49 raspberrypi4-64 avahi-daemon[379]: Server startup complete. Host name is raspberrypi4-64.local. Local service cookie is 322533974.
May 24 14:32:54 raspberrypi4-64 avahi-daemon[379]: Joining mDNS multicast group on interface wlan0.IPv6 with address fe80::c354:14d2:851e:ed1f.
May 24 14:32:54 raspberrypi4-64 avahi-daemon[379]: New relevant interface wlan0.IPv6 for mDNS.
May 24 14:32:54 raspberrypi4-64 avahi-daemon[379]: Registering new address record for fe80::c354:14d2:851e:ed1f on wlan0.*.
May 24 14:32:54 raspberrypi4-64 avahi-daemon[379]: Joining mDNS multicast group on interface wlan0.IPv4 with address 192.168.1.60.
May 24 14:32:54 raspberrypi4-64 avahi-daemon[379]: New relevant interface wlan0.IPv4 for mDNS.
May 24 14:32:54 raspberrypi4-64 avahi-daemon[379]: Registering new address record for 192.168.1.60 on wlan0.IPv4.
May 24 14:32:55 raspberrypi4-64 avahi-daemon[379]: Registering new address record for fe80::dea6:32ff:feb0:64e on wlan0.*.
May 24 14:33:01 raspberrypi4-64 avahi-daemon[379]: Registering new address record for 192.168.1.57 on wlan0.IPv4.
Nov 07 13:49:15 pn-v0-4-0 avahi-daemon[379]: Got SIGTERM, quitting.
Nov 07 13:49:15 pn-v0-4-0 avahi-daemon[379]: Leaving mDNS multicast group on interface wlan0.IPv6 with address fe80::c354:14d2:851e:ed1f.
Nov 07 13:49:15 pn-v0-4-0 avahi-daemon[379]: Leaving mDNS multicast group on interface wlan0.IPv4 with address 192.168.1.60.
Nov 07 13:49:15 pn-v0-4-0 avahi-daemon[379]: Leaving mDNS multicast group on interface lo.IPv6 with address ::1.
Nov 07 13:49:15 pn-v0-4-0 avahi-daemon[379]: Leaving mDNS multicast group on interface lo.IPv4 with address 127.0.0.1.
Nov 07 13:49:15 pn-v0-4-0 avahi-daemon[379]: avahi-daemon 0.8 exiting.
Nov 07 13:49:15 pn-v0-4-0 systemd[1]: Stopping Avahi mDNS/DNS-SD Stack...
Nov 07 13:49:15 pn-v0-4-0 systemd[1]: avahi-daemon.service: Deactivated successfully.
Nov 07 13:49:15 pn-v0-4-0 systemd[1]: Stopped Avahi mDNS/DNS-SD Stack.
Nov 07 13:49:15 pn-v0-4-0 systemd[1]: Starting Avahi mDNS/DNS-SD Stack...
Nov 07 13:49:15 pn-v0-4-0 avahi-daemon[639]: Found user 'avahi' (UID 994) and group 'avahi' (GID 993).
Nov 07 13:49:15 pn-v0-4-0 avahi-daemon[639]: Successfully dropped root privileges.
Nov 07 13:49:15 pn-v0-4-0 avahi-daemon[639]: avahi-daemon 0.8 starting up.
Nov 07 13:49:15 pn-v0-4-0 systemd[1]: Started Avahi mDNS/DNS-SD Stack.
Nov 07 13:49:15 pn-v0-4-0 avahi-daemon[639]: Successfully called chroot().
Nov 07 13:49:15 pn-v0-4-0 avahi-daemon[639]: Successfully dropped remaining capabilities.
Nov 07 13:49:15 pn-v0-4-0 avahi-daemon[639]: No service file found in /etc/avahi/services.
Nov 07 13:49:15 pn-v0-4-0 avahi-daemon[639]: Joining mDNS multicast group on interface wlan0.IPv6 with address fe80::dea6:32ff:feb0:64e.
Nov 07 13:49:15 pn-v0-4-0 avahi-daemon[639]: New relevant interface wlan0.IPv6 for mDNS.
```

**Comment by leigh-johnson** [comment]
:+1: Cool, I fixed this by always restarting systemd-resolvd -> avahi-daemon -> janus-gateway services after cloud-init has finished first boot run. Fix will be available in v0.4.3!
